### ISSUE TITLE
Not an issue, but a question: what in my settings is changing the html5 color picker popup?

### ISSUE THREAD

**Issue by prosoitos** [qutebrowser/qutebrowser#3849]
When selecting custom html5 colors on the site https://www.w3schools.com/colors/colors_picker.asp, here is what the popup window looks like in chromium:

![image](https://user-images.githubusercontent.com/4634851/38856475-6d3eefe4-41db-11e8-9a89-4a5af6bded92.png)

This is very convenient as I only have to move the gliders to a colour I like.

With qutebrowser `--temp-basedir`, here is what I get:

![image](https://user-images.githubusercontent.com/4634851/38857094-35803ca0-41dd-11e8-9060-79d101c4ca9f.png)

This is great and even better than the chromium one (though, I'd be curious to know why it is different).

With qutebrowser and my settings, here is what the popup window looks like:

![image](https://user-images.githubusercontent.com/4634851/38856519-8b4bf810-41db-11e8-9886-fd015577a0a0.png)

It is much less convenient because I first have to select "+" then paste the colour I had already pasted in the color value box above as it does not take this into account and finally the gliders are not all that practical.

Yesterday, while not changing any of my settings, I had the `--temp-basedir` popup. Back to the one I am familiar with (and don't like) today. Since I didn't change any setting at all, neither yesterday nor today, I imagine that yesterday my config file didn't load properly (computer needed a reboot, too many things running at the same time, that sort of thing) and whatever setting is affecting the default was not loaded properly.

I'd like to figure out what in my settings could be doing this and I'd be curious to understand a little bit why there are so many different popups depending on how I view this page... That may seem totally trivial, but I use that page very often.

Thank you so much for your time!

**Comment by The-Compiler** [comment]
Weird! I get the same you have in your middle picture with a PyQt from PyPI. But with my system-wide Qt, I get this, which is yet another one:

![image](https://user-images.githubusercontent.com/625793/38858112-2af9cea6-422b-11e8-9d9f-4e10bcca2780.png)

:laughing: 

Are you using the same Qt install (and version) in both cases? You can check the "Qt library executable path" in `:version` to find out.

**Comment by prosoitos** [comment]
Yes:

Whether I run qutebrowser with my settings or with `--temp-basedir`, here is what I have:
```
Qt library executable path: /usr/lib/qt/libexec, data path: /usr/share/qt
```

**Comment by prosoitos** [comment]
Only the `Paths` are different. Everything else is the same in both cases.

**Comment by toofar** [comment]
The middle one is shown as an example here too: http://doc.qt.io/archives/qt-5.5/qcolordialog.html
It's also what I see.

**Comment by prosoitos** [comment]
@The-Compiler : I launch both from the terminal (one with the `--temp-basedir` option and my regular one without (using rofi, but that shouldn't make a difference and I just tested it)).

**Comment by The-Compiler** [comment]
I'm out of ideas. Can you please do `:report` in both scenarios (and make sure to mention which is which) so I can check whether I can find some difference?

**Comment by prosoitos** [comment]
Done.

**Comment by The-Compiler** [comment]
I still suspect environment variables. Here's what's different with your regular config compared to `--temp-basedir`:

```
DESKTOP_SESSION = LXDE
PATH = /usr/local/sbin:/usr/local/bin:/usr/bin:/home/marie/bin:/usr/lib/jvm/default/bin:/usr/bin/site_perl:/usr/bin/vendor_perl:/usr/bin/core_perl
QT_AUTO_SCREEN_SCALE_FACTOR = 1
XDG_CONFIG_DIRS = /etc/xdg
XDG_CONFIG_HOME = /home/marie/.config
XDG_CURRENT_DESKTOP = LXDE
XDG_DATA_DIRS = /usr/local/share:/usr/share:/usr/share/gdm:/var/lib/menu-xdg
XDG_DATA_HOME = /home/marie/.local/share
XDG_MENU_PREFIX = lxde-
```

The ones I'd suspect are the culprit are `DESKTOP_SESSION` and `XDG_CURRENT_DESKTOP`, which maybe case LXDE's colorpicker (if such a thing exists) to be used.

I have no idea how they're set though, seeing you said you're launching from the same terminal? Did you make sure to quit qutebrowser before trying to launch it via the terminal?

**Comment by prosoitos** [comment]
Also, when qutebrowser opens from another application (with `mimeo` opening it in the background in my settings), I also have the crappy color picker in that session.

**Comment by slackhead** [comment]
I've noticed things like this with rofi and environmental vars in the past and have had to launch it with eg:
PATH=$HOME/scripts:$HOME/bin:$PATH rofi

To get my normal $PATH to show. The extra ones were set in my ~/.zshrc and don't seem to propagate to rofi (in i3 at least). It seems ok in fluxbox, so perhaps it's an i3 problem.

**Comment by prosoitos** [comment]
Thank you for the comments @slackhead. I will check whether this applies to me!

**Comment by slackhead** [comment]
There is something you could test:

Using rofi, open a terminal and run 'env > ~/rofi-env.txt' (without the quotes).

Then open a termninal without using rofi and run 'env > ~/non-rofi-env.txt'

Then it's a matter of comparing the two files to see what's different, if anything. The 'env' command will list all environmental variables. You could try using 'diff' on them:

diff rofi-env.txt non-rofi-env.txt

Maybe something will show up, but it can depend on the order that things are listed.

Hope this helps.

**Comment by The-Compiler** [comment]
I digged into QtWebEngine's sources a bit, and it seems to use a QML [ColorDialog](https://doc.qt.io/qt-5/qml-qtquick-dialogs-colordialog.html). From the docs for it:

> The implementation of ColorDialog will be a platform color dialog if possible. If that isn't possible, then it will try to instantiate a QColorDialog. If that also isn't possible, then it will fall back to a QML implementation

The good dialog you get is indeed a [QColorDialog](https://doc.qt.io/qt-5/qcolordialog.html) (see the screenshot there), so it looks like ColorDialog is indeed somehow showing an LXDE color dialog.

(Chromium's is different because it has its own color chooser widget, which Qt doesn't use at all)

**Comment by The-Compiler** [comment]
Out of curiosity: Can you run `:spawn -o qtdiag` in qutebrowser in both scenarios, and show the "Theme:" part of it?

**Comment by prosoitos** [comment]
@slackhead: thanks for your time!

Finally had a bit of time to follow your advice:

- `PATH` shows up as identical in the environment, whether the called directly from `lxterminal` (equivalent to xterm that I use in my system) or from `lxterminal` used by rofi.

In fact, everything in the environment is the same with the exception that `lxterminal`, when opened directly, is at the 2nd level deep (normal): `SHLVL=2` and when it is called by rofi, it is at the 3rd level deep, which, I guess, might make sense: `SHLVL=3`. I don't know if this may matter for the issue here.

Note: things are listed in the exact same order except for `_=/usr/bin/env` (shows up in both, but at a different place).

**Comment by prosoitos** [comment]
@The-Compiler: thanks for your research. Things make a lot more sense already.

> Out of curiosity: Can you run :spawn -o qtdiag in qutebrowser in both scenarios (part of qt5-tools), and show the "Theme:" part of it?

`Error while spawning command: The process failed to start.` (in both cases) :confused:

**Comment by The-Compiler** [comment]
Do you have `qt5-tools` installed? If you do, you should be able to debug this by running `qtdiag` in a terminal first.

**Comment by prosoitos** [comment]
Thanks. I am so not a qt person...

Good intuition: the themes are different:

- qutebrowser opened from terminal:
```
Theme:
  Platforms requested : unknown,generic
            available : gtk3
  Styles requested    : Fusion,Windows
         available    : Windows,Fusion
  System font         : "Sans Serif" 9
```
- qutebrowser opened by rofi:
```
Theme:
  Platforms requested : gtk3,gnome,generic
            available : gtk3
  Styles requested    : fusion,windows
         available    : Windows,Fusion
  Icon theme          : Adwaita,  from /home/marie/.icons,/usr/share/icons,/usr/share/pixmaps
  System font         : "Sans" 10
  Native file dialog
  Native color dialog
  Native font dialog
```

**Comment by The-Compiler** [comment]
You do have a `config.py` IIRC, right? What happens if you do something like this there?

```python
os.environ.pop('DESKTOP_SESSION')
os.environ.pop('XDG_CURRENT_DESKTOP')
```

**Comment by The-Compiler** [comment]
Going to close this one now, as this is all I can say really - I have no clue how Qt selects its platform plugins if it's not via those environment variables. I'm glad we could at least figure it out partially, though! :smile:

**Comment by prosoitos** [comment]
Yes, thank you! This was very helpful: at least, now I know how to launch qutebrowser to have the better popup window!

**Comment by The-Compiler** [comment]
FWIW I just learned that there's a `QT_QPA_PLATFORMTHEME` environment variable - I tried doing `os.environ['QT_QPA_PLATFORMTHEME'] = 'kde'` and indeed got a different color picker (and e.g. file chooser), so you might want to experiment with that. The platformthemes are in `/usr/lib/qt/plugins/platformthemes/` but I have no idea how to get the valid names to use.

**Comment by prosoitos** [comment]
Oh, interesting! Do you know what to set `os.environ['QT_QPA_PLATFORMTHEME']` to to get the default color picker?

**Comment by The-Compiler** [comment]
No clue - maybe just try setting it to an invalid value (`'foo'` or whatever)
